### DISCUSSION TITLE
Error Handling Design

### DISCUSSION THREAD

**Discussion by lmverity** [ecederstrand/exchangelib#875] [Q&A]
Hi!
I've put in a few PRs for error handling changes; alas, we've had another instance where we can't handle the errors coming back from EWS, since they're mostly being raised in the response in exchangelib. For context, our design approach requires processing all items in a mailbox, and skipping those that cause errors due to things like corrupt data, permissions issues, etc. We've mostly been able to work around not receiving some errors at the app level, but it's getting more difficult as we scale to work in different environments and encounter items with a wider variety of issues.
I was wondering if you would mind sharing the thought process behind raising many of the errors in the response, rather than returning them to be handled at some higher level in the apps that are using exchangelib. We're struggling with how to integrate our approach with the current error handling setup, and it would help to have more context around the design choices that were made there. 
I appreciate all your work on this project--I work with exchangelib every day now and have learned quite a bit from it. :)
Thanks!

**Comment by ecederstrand** [comment]
The general idea with not returning all exceptions to the user is to only return exceptions that are understood. Different exceptions require different actions. For example, some exceptions require us to throttle the connections, and I'd like to not leave that up to the user.

There are two levels of exceptions in EWS - the general SOAP exceptions that concern the entire response, and per-item exceptions that concern only the individual item being returned. Mostly, the per-item exceptions can be returned to the user, while the general SOAP exceptions should be raised.

The short-term solution for you would be to add the exceptions you prefer to pass on to the user to the `ERRORS_TO_CATCH_IN_RESPONSE` attribute. You can do that in your own code, but I'd like to hear about the ones you do add so we can consider adding them to this package.

I'd also like to hear your thoughts on the above.

**Reply by lmverity** [reply]
That makes sense! We can certainly pull out per-item exceptions that make sense to include in `ERRORS_TO_CATCH_IN_RESPONSE` and add them to our code. I wonder if it's worth the time for me to go through the whole list of EWS exceptions from the MS documentation and try to pull out any of the errors that are per-item; for us, that would guarantee, or at least add additional confidence, around per-item errors not breaking due to EWS exceptions we otherwise could handle. Would that be useful as a contribution as well?

**Reply by ecederstrand** [reply]
Yes, that would be very useful. Theres a list of errors and descriptions at https://docs.microsoft.com/en-us/exchange/client-developer/web-service-reference/responsecode but it's not clear to me which errors belong in which category.

I've been using exchangelib for syncing with a range of servers the last 5-10 years, so the list of known errors covers the vast majority, but I'll never see the full range of failure scenarios that the combined users of exchangelib will stumble into.

**Reply by lmverity** [reply]
Awesome, I'll take a look at it and start some categorizing!
